### PR TITLE
feat: ECharts Phase 2 - Custom ECharts

### LINKED ISSUE #24424: [Task]: Add an option to render custom ECharts option in chart type dropdown
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

1. Users should be able to select custom echarts option from chart type dropdown.
2. The custom fusion chart should appear as a deprecated option and that it will be phased out soon.

### LINKED ISSUE #25564: [Task]: Add deprecation UI for custom fusion charts
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

Need to add UI for this figma : https://www.figma.com/file/yLnYf2QslxjgSKbsIWrs4w/Fusion-Charts--%3E-E-Charts?type=design&node-id=105%3A13555&mode=design&t=KjIq52iIBwASc354-1

### LINKED ISSUE #26009: [Task]: Write a data migration to add default custom echart configuration for existing chart widgets
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

The existing chart widgets on prod should have default custom echart data prefilled for them when custom echart gets released.

### LINKED ISSUE #26545: [Task]: Add feature flag for custom echarts and custom fusion chart deprecation message
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

We would like to have custom echart and custom fusion chart deprecation message behind feature flags.

### LINKED ISSUE #26584: [Task]: Add support for rawEventData in custom ECharts
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

Clicking on a custom echart should pass correct rawEventData to the user.

### PR DESCRIPTION
**Pull request by rajatagrawal**
Fixes #24424
Fixes #26009
Fixes #25564
Fixes #26545
Fixes #26584

### CODE REVIEW

**Comment by chandannkumar** [comment]
@rajatagrawal 
Observations (as discussed):
- Widget breaks on emptying chart datasource/configuration. Also break on adding incorrect datasource
[sample echart.txt](https://github.com/appsmithorg/appsmith/files/12278989/sample.echart.txt)
https://www.loom.com/share/2d2045a0b9344d87854d0269d923c245?sid=6d393b4f-9dad-4e6d-b234-8980ffeefe81

- Merge the Custom Echarts Datasource & configuration property into 1
image

- Tooltip of custom Echart configuration seems quite big with extra spaces
image

File: app/client/src/widgets/ChartWidget/component/EChartsDatasetBuilder.ts [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
-  static chartData(props: ChartComponentProps): AllChartData {
-    if (props.chartType == "PIE_CHART") {
-      const firstKey = Object.keys(props.chartData)[0];
-      return { [firstKey]: props.chartData[firstKey] };
+  static chartData(
+    chartType: ChartType,
+    chartData: AllChartData,
+  ): AllChartData {
+    if (chartType == "PIE_CHART") {
+      const firstKey = Object.keys(chartData)[0];
```
Comment (sbalaji1192) [inline app/client/src/widgets/ChartWidget/component/EChartsDatasetBuilder.ts:10]: Can you add a brief comment of what we are doing here?

File: app/client/src/widgets/ChartWidget/widget/SyntaxErrorsEvaluation.ts [hunk 2]
```diff
@@ -1,55 +1,70 @@
...
-          if (props.chartType == "PIE_CHART") {
-            const firstSeriesKey = Object.keys(props.chartData)[0];
-            if (
-              error.path &&
-              error.path.startsWith(`chartData.${firstSeriesKey}`)
-            ) {
-              errors.push(error);
-            }
-          } else {
+      if (
```
Comment (sbalaji1192) [inline app/client/src/widgets/ChartWidget/widget/SyntaxErrorsEvaluation.ts:16]: can we club these conditions, Looks like we're pushing to errors in all cases.

Comment (sbalaji1192) [reply app/client/src/widgets/ChartWidget/widget/SyntaxErrorsEvaluation.ts:16]: Can we create a better strucuter for this error checking? I feel if else bit messy here

File: app/client/src/utils/migrations/ChartWidget.ts [hunk 3]
```diff
@@ -16,3 +17,12 @@ export const migrateChartWidgetLabelOrientationStaggerOption = (
     }
   });
 };
+
+export const migrateDefaultValuesForCustomEChart = (currentDSL: DSLWidget) => {
+  return traverseDSLAndMigrate(currentDSL, (widget: WidgetProps) => {
+    if (widget.type == "CHART_WIDGET") {
+      const chartWidgetProps = widget as ChartWidgetProps;
+      chartWidgetProps.customEChartConfig = DefaultEChartConfig;
```
Comment (sbalaji1192) [inline app/client/src/utils/migrations/ChartWidget.ts:25]: We should not import constants and use it migrations. `DefaultEChartConfig` value is going to change as the widget evolves. We should use literal values so this migration will always refer to same value.

File: app/client/src/widgets/ChartWidget/component/EChartsDatasetBuilder.ts (hunk 1 above)
Comment (rajatagrawal) [reply app/client/src/widgets/ChartWidget/component/EChartsDatasetBuilder.ts:10]: Done

File: app/client/src/utils/migrations/ChartWidget.ts (hunk 3 above)
Comment (rajatagrawal) [reply app/client/src/utils/migrations/ChartWidget.ts:34]: Done

File: app/client/src/widgets/ChartWidget/widget/propertyConfig.ts [hunk 4]
```diff
@@ -284,7 +310,9 @@ export const contentConfig = [
         isBindProperty: true,
         isTriggerProperty: false,
         validation: { type: ValidationTypes.TEXT },
-        hidden: (x: any) => x.chartType === "CUSTOM_FUSION_CHART",
+        hidden: (x: any) =>
+          x.chartType === "CUSTOM_FUSION_CHART" ||
```
[the remark this answers is not in the document]
Comment (rajatagrawal) [reply app/client/src/widgets/ChartWidget/widget/propertyConfig.ts]: Done

File: app/client/src/widgets/ChartWidget/widget/propertyConfig.ts [hunk 5]
```diff
@@ -295,7 +333,8 @@ export const contentConfig = [
         isBindProperty: true,
         isTriggerProperty: false,
         validation: { type: ValidationTypes.TEXT },
-        hidden: (x: any) => x.chartType === "CUSTOM_FUSION_CHART",
+        hidden: (x: any) =>
```
Comment (sbalaji1192) [inline app/client/src/widgets/ChartWidget/widget/propertyConfig.ts:336]: lets use proper argument name like , widget or props and use appropriate type

File: app/client/src/widgets/ChartWidget/constants.ts [hunk 6]
```diff
@@ -41,6 +45,8 @@ export const messages = {
   Undefined: "Undefined",
 };
 
+export const isCustomEChartEnabled = false; // FEATURE_FLAG["release_enable_custom_echarts"]
```
[the remark this answers is not in the document]
Comment (rajatagrawal) [reply app/client/src/widgets/ChartWidget/constants.ts]: Yes, I have removed this statement now.

File: app/client/src/widgets/ChartWidget/widget/propertyConfig.ts (hunk 5 above)
Comment (rajatagrawal) [reply app/client/src/widgets/ChartWidget/widget/propertyConfig.ts]: Done

**Comment by laveena-en** [comment]
@rajatagrawal A couple of callouts mentioned below:

3D charts aren’t being rendered completely: https://www.loom.com/share/d85df09429da4abcb77bcc6f655b048a?sid=64445193-2384-49db-93fc-9e553fe57e8e

Legends are overlapping when the chart is resized. However, Fusion charts introduce arrows to navigate between legends when the size of the chart is small
image

Default custom E-charts need to be changed to a different type as discussed

The widget just goes blank without throwing an error when fusion chart data is added into E-charts confighttps://www.loom.com/share/6dd4e58cabd440eaa2ce469e3a5bcf94?sid=26548278-d542-403e-8ec3-c2464028451e

Chart data isn’t reflecting the E-charts config changes: https://www.loom.com/share/c8589574547145baa43f97a966553268?sid=bc276be0-acbc-414f-9fa7-597068ae4f46

**Review by hetunandu — COMMENTED** [review]
FF changes LGTM

File: app/client/src/pages/Editor/PropertyPane/PropertyPaneView.tsx [hunk 7]
```diff
@@ -261,6 +259,7 @@ function PropertyPaneView(
             {deprecationMessage}
           </Callout>
         )}
+        {editorCallouts(widgetProperties)}
```
Comment (sbalaji1192) [inline app/client/src/pages/Editor/PropertyPane/PropertyPaneView.tsx:262]: lets call this renderWidgetCallouts

File: app/client/src/widgets/ChartWidget/widget/index.tsx [hunk 8]
```diff
@@ -79,6 +109,31 @@ class ChartWidget extends BaseWidget<ChartWidgetProps, WidgetState> {
     };
   }
 
+  static editorCallouts() {
+    const callouts = (props: WidgetProps): WidgetCallout[] => {
```
[the remark this answers is not in the document]
Comment (rajatagrawal) [reply app/client/src/widgets/ChartWidget/widget/index.tsx:113]: We are calling this function in chart widget's configuration index file in getWidgetMethods field which accepts a function. Property pane needs this function to pass props to it and get relevant callouts.

[the remark this answers is not in the document]
Comment (sbalaji1192) [reply app/client/src/widgets/ChartWidget/widget/index.tsx:113]: We can assign the function to the methods object, right? why we need to call and return a function

[the remark this answers is not in the document]
Comment (rajatagrawal) [reply app/client/src/widgets/ChartWidget/widget/index.tsx:113]: Done

**Review by sbalaji1192 — APPROVED** [review]

**Pull request merged**